### PR TITLE
Make String a must_use type

### PR DESCRIPTION
**Pull request by tesuji**
Make String a must_use type to lint again unused String: #75742

### CODE REVIEW

File: src/librustdoc/formats/cache.rs [hunk 1]
```diff
@@ -472,7 +472,7 @@ impl DocFolder for Cache {
         });
 
         if pushed {
-            self.stack.pop().expect("stack already empty");
+            let _ = self.stack.pop().expect("stack already empty");
```
Comment (scottmcm) [inline src/librustdoc/formats/cache.rs:475]: I think lines like this are a good example of why `String` and `Vec<T>` aren't `must_use`.

Comment (lzutao) [reply src/librustdoc/formats/cache.rs:475]: These cases are rare in Rustc codebase.

Comment (CryZe) [reply src/librustdoc/formats/cache.rs:475]: I agree with @scottmcm here. If it's on String, then it should be on Vec too, and if it's on Vec, then it should also be on every container. At which point you'd have to somewhat argue why it's not on basically everything that's not Copy.

Comment (lzutao) [reply src/librustdoc/formats/cache.rs:475]: With all respects, I just want some actions on resolving #75742 . If the team doesn't like this approach,
what about marking `String::from`, `str::into`, `fmt::format` with `must_use` attr?

**Review by dtolnay — COMMENTED** [review]
I agree with @scottmcm above. A String being present in a return value does not mean it's a newly allocated string that you could hope to optimize away when unused, as implied by the PR's must_use message of `"unused allocation might not be optimized out by compiler"`. I am concerned that must_use has been trending noisier and noisier as it appears in increasingly questionable places and I would prefer to be more cautious and attentive about false positives when applying it.

**Comment by dtolnay** [comment]
> With all respects, I just want some actions on resolving #75742 . If the team doesn't like this approach,
what about marking `String::from`, `str::into`, `fmt::format` with `must_use` attr?

I don't see the connection of those suggestions toward resolving #75742. That issue is motivated by `f(&format!(...))` where `f` is inlinable and does not access its argument. As far as must_use would be concerned it's still a use of the value because the value gets bound to a variable (the function's formal argument).

Independent of #75742, I would still recommend against going for must_use on those functions unless there is some evidence of pattern of misuse of those functions that we are hoping to prevent. Aggressively tagging must_use without a pattern of misuse just increases the false positive rate and general nuisance of must_use without delivering practical benefit.

**Comment by lzutao** [comment]
> I don't see the connection of those suggestions toward resolving #75742.

I meant: `format!` expands to `fmt::format(Arguments)`, so marking `fmt::format` result `must_use`
will make people be aware of unused allocated String. So they will use the String or remove the useless expression
or replace it with `to_string` method or so.
I haven't ever seen the meaning of unused `fmt::format` result.

But yeah, if the lib team has other ways to resolve the issue, I am happy to wait for it.

**Comment by scottmcm** [comment]
@lzutao An idea you might be interested to play with: the [`unused_results` lint](https://doc.rust-lang.org/rustc/lints/listing/allowed-by-default.html#unused-results) is allow-by-default because it's so noisy.  Maybe there's some subset of it -- perhaps "unused result where the unused type happens to have drop glue and came from a `&impl Freeze` method", but exact rules obviously TBD -- that could be worth experimenting with to see if there's a version of it that people would be more inclined to actually turn on.

**Pull request closed without merging**